### DISCUSSION TITLE
External variables are probably too burdensome

### DISCUSSION THREAD

**Discussion by AntoineThebaud** [perses/perses#1545] [Design Discussions]
I think that in their current state, global variables with display.hidden set to false are never going to be used, because adding such variable implies that a new dropdown will show up on each and every dashboard saved in Perses (unless locally overriden), which is a strong assumption that it's going to make sense in all of them.

I think this assumption is unrealistic, and the current way of working around it (aka override with a dashboard variable & put its visibility to hidden) is not convenient at all. Imagine if you have 5 global variables defined and only 1 of them makes sense on your dashboard..

So I'd say, wait & see to have more real usage data before taking any action here, but I foresee that (visible) global variables may be never used because of this.

To me the solution here is to be able to toggle the global variables visibility at dashboard level, by storing a reference to them. This way you could still reference them here and there in the dashboard definition, but the end users of your dashboard wouldn't see any dropdown generated for it.
A limitation we could think of is that, in case a given global variable would be removed or renamed (not possible for now but could be at some point), dashboard designers would have to update/redo the variable reference accordingly. However it's actually not a new issue since currently if you do so, all dashboards relying on this global variable e.g in their panel queries will have them all broken, until they update all the definitions to use the new name.

**Comment by Nexucis** [comment]
and what about only display variable that are actually used in the dashboard ? Regardless where does it come from

**Reply by AntoineThebaud** [reply]
Interesting, would you check the entire dashboard content or only the local variables?

**Reply by Nexucis** [reply]
I would check the entire dashboard

**Comment by AntoineThebaud** [comment]
Just thinking about this again and another issue our current "external" variables involve: imagine that you have
- a global variable that provides the full list of your kubernetes clusters,
- on half of your dashboards you want to have multi-selection of values available,
- on the other half you don't.

-> In that case, you would have to choose one of the two options for the global var, use it as-is on the first half of dashboards & override it on the second, i.e define a local variable on each dashboard just to change that single multi-selection field value. Thus if ever in the future the query to get the list of clusters has to change, updating the global variable won't affect the second half of dashboards, so you'll have to update a local variable on each one of them.

That feels very cumbersome to me... So to solve this (+ the other issue I mentioned in the original post), I think we should somehow provide (global/project) variables "templates" instead, with local variables that would be linked to a given template, inherit what's define in it & define values locally only for the fields where the base template values don't fit. "Linked" involves that when the template gets updated, this update would get automatically propagated to all variables associated to it (maybe not simple to implement, I think we already discussed such behavior for the "visibility" mentioned in original post)

This way we'd keep providing the variables factorization feature we wanted when we designed "external" variables, while getting rid of the limitations / strong assumptions these ended up to imply.

Btw well-voted Grafana issue that seem to correspond to this idea https://github.com/grafana/grafana/issues/1959.

**Comment by AntoineThebaud** [comment]
Another problem with external variables that we already face and is quite annoying: you cannot save default values for those at dashboard level.
